### PR TITLE
update requests to allow bound-withs

### LINKED ISSUE #1446: Title level: display explanatory info about Bound-with parent
So that a user isn't confused when they ultimately receive the parent item. The user should expect to receive the parent item with their requested child item bound inside.

See https://github.com/sul-dlss/SearchWorks/issues/3200 for more context

Could test this out by sending a query param, even without editing the button from SearchWorks.

This is the view:

Screenshot 2024-01-31 at 5 07 04 PM

### PR DESCRIPTION
**Pull request by dnoneill**
closes #1446 

**I was not able to test actually creating a request locally. I am pretty sure based on the code that call number isn't something get sent for single items but I could not confirm that**
What this PR does:
1. updates instance graphql query to query `holdingsRecord` for `boundWithItem` records for those holdings. This gets parent_bound_withs for the record. It also adds `boundWithHoldingsPerItem` to the item query that is used both by the `boundWithItem` and `instance`.
2. Updates folio instance model to get items not only from the response but also from the boundWithItem field. This can cause an increase in items (see http://localhost:3000/pages/new?item_id=387559&origin=SAL3&origin_location=SAL3-STACKS). This example adds the [A 1.34:87 bound with A 1.34:81. 36105019614499 (item id)](https://searchworks.stanford.edu/view/10564224) 
3. This ticket also adds a list of bound-withs for an item if they have them.
4. Moves call number

Note: after discussion with Darcy the call number/title display is in one format for maintenance considerations.

1. http://localhost:3000/hold_recalls/new?item_id=6290245&origin=GREEN&origin_location=GRE-STACKS
Screenshot 2024-03-06 at 11 39 24 AM

4. http://localhost:3000/hold_recalls/new?barcode=36105029307795&item_id=6290236&origin=GREEN&origin_location=GRE-STACKS
Screenshot 2024-03-06 at 11 38 51 AM

5. http://localhost:3000/pages/new?item_id=2969234&origin=SAL3&origin_location=SAL3-STACKS
Screenshot 2024-03-06 at 11 18 24 AM

6. http://localhost:3000/pages/new?item_id=10564224&origin=SAL3&origin_location=SAL3-STACKS

Screenshot 2024-03-06 at 12 25 54 PM

### CODE REVIEW

File: app/assets/stylesheets/modules/forms.scss [hunk 1]
```diff
@@ -65,6 +65,11 @@ hr {
   white-space: nowrap;
 }
 
+.single-item-callnumber {
+  margin: -1em 0 0.5em;
+  color: #8c1515;
```
Comment (jcoyne) [inline app/assets/stylesheets/modules/forms.scss:70]: Please make this a variable and put it in https://github.com/sul-dlss/sul-requests/blob/main/app/assets/stylesheets/sul-variables.scss#L1

File: app/assets/stylesheets/modules/item-selector.scss [hunk 2]
```diff
@@ -86,3 +86,34 @@
   }
 }
 
+.bound-withs {
+  max-height: 200px;
+  background-color: #F4F4F4;
```
Comment (jcoyne) [inline app/assets/stylesheets/modules/item-selector.scss:91]: Please make this a variable and put it in https://github.com/sul-dlss/sul-requests/blob/main/app/assets/stylesheets/sul-variables.scss#L1

File: app/assets/stylesheets/modules/item-selector.scss [hunk 3]
```diff
@@ -86,3 +86,34 @@
...
+  overflow: scroll;
+  padding: 8px;
+  margin-bottom: 15px;
+  border-radius: 4px;
+
+  .bound-with-item span{
```
Comment (jcoyne) [inline app/assets/stylesheets/modules/item-selector.scss:97]: Can we just use a `div` instead?

File: app/assets/stylesheets/modules/item-selector.scss [hunk 4]
```diff
@@ -86,3 +86,34 @@
...
+    padding: 4px 0px;
+  }
+
+  .heading {
+    font-weight: 500;
+    margin-bottom: 10px;
+  }
+
+  li:not(:only-child) {
+    border-top: 1px solid #D5D5D4;
```
Comment (jcoyne) [inline app/assets/stylesheets/modules/item-selector.scss:111]: Please make this color a variable and put it in https://github.com/sul-dlss/sul-requests/blob/main/app/assets/stylesheets/sul-variables.scss#L1

File: app/models/folio/instance.rb [hunk 5]
```diff
@@ -14,6 +14,44 @@ def self.fetch(request, _live_lookup = true)
...
+    end
+
+    def self.update_item_fields(item, holdings_record)
+      item['effectiveCallNumberComponents']['callNumber'] = holdings_record['callNumber']
+      item['volume'] = ''
+      item['enumeration'] = ''
+      item
+    end
+
+    def self.get_items(json)
```
Comment (jcoyne) [inline app/models/folio/instance.rb:37]: Using `get_` is sort of frowned on in ruby. Can we use `items(json)` instead?  The "getting" is implied.

File: app/models/folio/instance.rb [hunk 6]
```diff
@@ -14,6 +14,44 @@ def self.fetch(request, _live_lookup = true)
...
+      return json.fetch('items', []).map { |item| Folio::Item.from_hash(item) } unless parent_bound_withs.present? && hrid
+
+      parent_bound_withs = parent_bound_withs.pluck('boundWithItem')
+      items = filter_parent_bound_withs(parent_bound_withs, hrid)
+      items += json['items'] if json['items']
+      items.compact.map { |item| Folio::Item.from_hash(item) }
+    end
+
+    def self.parent_bound_withs(json)
+      json.fetch('holdingsRecords', [])&.filter { |holding| holding['boundWithItem'] }
```
Comment (jcoyne) [inline app/models/folio/instance.rb:49]: Is the safe navigation necessary here?

File: app/models/folio/instance.rb [hunk 7]
```diff
@@ -14,6 +14,44 @@ def self.fetch(request, _live_lookup = true)
...
+    end
+
+    def self.child_bound_withs(json)
+      json['items']&.pluck('boundWithHoldingsPerItem')&.flatten&.compact
```
Comment (jcoyne) [inline app/models/folio/instance.rb:53]: If you do `return unless json['items']`, then you don't need all of these safe navigation operators.

File: app/views/application/_item_selector.html.erb [hunk 8]
```diff
@@ -2,20 +2,26 @@
...
+      <div class="single-item-callnumber">
+        Call number: <%= single_item.callnumber %>
       </div>
       <%= barcode.hidden_field single_item.barcode.present? ? single_item.barcode : 'NO_BARCODE', value: 1 %>
+      <% if f.object.bound_with? %>
+        <div class="bound-withs">
+            <% is_child = f.object.parent_bound_withs.present? %>
+            <div class="heading">This item is bound<% if is_child %> and shelved<% end %> with</div>
+            <% if is_child %>
+              <%= render partial: 'item_selector_bound_withs', locals: { bound_withs: f.object.parent_bound_withs, is_child: } %>
```
Comment (jcoyne) [inline app/views/application/_item_selector.html.erb:14]:
```suggestion
              <%= render 'item_selector_bound_withs', bound_withs: f.object.parent_bound_withs, is_child: %>
```

File: app/views/application/_item_selector.html.erb [hunk 9]
```diff
@@ -2,20 +2,26 @@
...
+            <% else %>
+              <%= render partial: 'item_selector_bound_withs', locals: { bound_withs: f.object.child_bound_withs, is_child: } %>
```
Comment (jcoyne) [inline app/views/application/_item_selector.html.erb:16]:
```suggestion
              <%= render 'item_selector_bound_withs',  bound_withs: f.object.child_bound_withs, is_child: %>
```

File: app/views/application/_item_selector.html.erb [hunk 10]
```diff
@@ -2,20 +2,26 @@
...
+            <% end%>
```
Comment (jcoyne) [inline app/views/application/_item_selector.html.erb:17]:
```suggestion
            <% end %>
```

File: app/views/application/_item_selector.html.erb [hunk 11]
```diff
@@ -2,20 +2,26 @@
...
+        Call number: <%= single_item.callnumber %>
       </div>
       <%= barcode.hidden_field single_item.barcode.present? ? single_item.barcode : 'NO_BARCODE', value: 1 %>
+      <% if f.object.bound_with? %>
+        <div class="bound-withs">
+            <% is_child = f.object.parent_bound_withs.present? %>
+            <div class="heading">This item is bound<% if is_child %> and shelved<% end %> with</div>
+            <% if is_child %>
+              <%= render partial: 'item_selector_bound_withs', locals: { bound_withs: f.object.parent_bound_withs, is_child: } %>
+            <% else %>
```
Comment (jcoyne) [inline app/views/application/_item_selector.html.erb:15]: This branch looks identical to the one above.

File: app/assets/stylesheets/modules/item-selector.scss [hunk 12]
```diff
@@ -86,3 +86,34 @@
   }
 }
 
+.bound-withs {
+  max-height: 200px;
```
Comment (jcoyne) [inline app/assets/stylesheets/modules/item-selector.scss:90]: Can we use relative units (rem or em) where appropriate?  This makes it easier to style for responsiveness to different screen sizes.

File: app/views/application/_item_selector_bound_withs.html.erb [hunk 13]
```diff
@@ -0,0 +1,10 @@
+<ul>
```
Comment (jcoyne) [inline app/views/application/_item_selector_bound_withs.html.erb:1]:
```suggestion
<ul  class="list-unstyled">
```

File: app/assets/stylesheets/modules/item-selector.scss [hunk 14]
```diff
@@ -86,3 +86,34 @@
...
+  .heading {
+    font-weight: 500;
+    margin-bottom: 10px;
+  }
+
+  li:not(:only-child) {
+    border-top: 1px solid #D5D5D4;
+  }
+
+  ul {
```
Comment (jcoyne) [inline app/assets/stylesheets/modules/item-selector.scss:114]: Can we use the Bootstrap  class="list-unstyled" instead?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/models/folio/instance.rb (hunk 6 above)
Comment (dnoneill) [reply app/models/folio/instance.rb:49]: I was thinking holdingsRecord can return nil but I am not sure that exists so no.

File: app/views/application/_item_selector.html.erb (hunk 11 above)
Comment (dnoneill) [reply app/views/application/_item_selector.html.erb:15]: I am not sure I understand what you are saying here.

Comment (dnoneill) [reply app/views/application/_item_selector.html.erb:15]: I am not sure I understand what you are saying.

**Review by jcoyne — APPROVED** [review]

**Comment by cbeer** [comment]
This is great. I've tested it out locally with some bound-with children + bound-with parents. It's great we're able to preserve the bound-with child information through mylibrary by associating the request with the bound-with child. I think we should verify Caiasoft is ok with that (Darsi is following up with them), but I don't think that testing needs to hold up this PR.

File: app/models/folio/instance.rb [hunk 15]
```diff
@@ -14,6 +14,46 @@ def self.fetch(request, _live_lookup = true)
...
+    def self.filter_parent_bound_withs(parent_bound_withs, hrid)
+      return [] unless parent_bound_withs
+
+      parent_bound_withs.flatten.compact.each do |bound_with_record|
+        @filter_parent_bound_withs ||= bound_with_record.dig('instance', 'items').each.map do |item|
+          boundwithhrids = item['boundWithHoldingsPerItem'].filter { |bw| bw.dig('instance', 'hrid') == hrid }.flatten
+          update_item_fields(item, boundwithhrids.first) if boundwithhrids.present?
+        end
+      end
+      @filter_parent_bound_withs.compact
```
Comment (cbeer) [inline app/models/folio/instance.rb:27]:
```suggestion
      parent_bound_withs.flatten.compact.each do |bound_with_record|
        filter_parent_bound_withs ||= bound_with_record.dig('instance', 'items').each.map do |item|
          boundwithhrids = item['boundWithHoldingsPerItem'].filter { |bw| bw.dig('instance', 'hrid') == hrid }.flatten
          update_item_fields(item, boundwithhrids.first) if boundwithhrids.present?
        end
      end
      filter_parent_bound_withs.compact
```

I don't think we want to memoize these at the class level here.

Comment (dnoneill) [reply app/models/folio/instance.rb:27]: @cbeer It breaks the application if we remove the memoization.

**Comment by dnoneill** [comment]
There is one thing that isn't totally consistent across pages. If it is a new page http://localhost:3000/pages/new?item_id=14029321&origin=SAL3&origin_location=SAL3-STACKS the call number will have more spacing because of the line https://github.com/sul-dlss/sul-requests/blob/7cb6f9b6ab04a2da1f78705474ec1d34d2357fc9/app/assets/stylesheets/modules/forms.scss#L15. I don't think it is a change that will stop this from being merged but I wanted to make sure it was documented.

**Pull request closed without merging**